### ISSUE TITLE
Execute when filter is enabled

### ISSUE THREAD

**Issue by tatunts** [upgradeQ/obs-libre-macros#2]
Hello sir!

This is a powerful script that could help a lot.

I was wondering: is it possible to execute the code the moment the filter is enabled? Also, when it finishes, disable the filter again.

If so, can you provide an example of a simple hotkey press when the filter is enabled, pass the hotkey, wait a bit then disable the filter.

Thanks!

**Comment by upgradeQ** [comment]
```lua
local state_of_filter = "task_name_state"
if not t.tasks[state_of_filter] then
  t.tasks[state_of_filter] = run(function()
    repeat sleep(0)
    if obs_source_enabled(t.filter) then
      print([[it's enabled]])
      print_source_name(source)
      sleep(2)
      print_source_name(t.filter)
      sleep(1)
      print('done')
      obs_source_set_enabled(t.filter, false)
    end
    until false
  end)
end

obs_source_set_enabled(t.filter, true)
```
There is also a [`source signals`](https://obsproject.com/docs/reference-sources.html#source-signals) which you could try, they are executed immediately.

**Comment by tatunts** [comment]
Thanks a lot! I’ll give it a try and let you know.

**Comment by tatunts** [comment]
Hello!

I've tried your code and it works, but the problem is that it needs to be executed at least once to "respond" to the filter state changes. I know there's a auto run check box that can help with that but I plan on having this code in about 20 separated filters, you can imagine OBS triggering all 20 actions at the same when it starts haha

Am I doing something wrong here? Is there a solution for that? Maybe set the code to just "be ready" for the filter state but do not trigger the hotkey when auto run is on and OBS starts...

Thanks!

**Comment by upgradeQ** [comment]
@tatunts which hotkey? It does not trigger code if last statement is omitted, with `Auto run` **on** it will schedule to execute itself in the future.

**Comment by tatunts** [comment]
Hello!

Sorry, I'm not very good at code :)

I've changed your previous code to add a send hotkey, like I've described before. I need to send a hotkey press to OBS when the console filter is on. This is the code I've tested:

```
local state_of_filter = "task_name_state"
if not t.tasks[state_of_filter] then
  t.tasks[state_of_filter] = run(function()
    repeat sleep(0)
    if obs_source_enabled(t.filter) then
      print([[it's enabled]])
      print_source_name(source)
      send_hotkey('OBS_KEY_F13')
      print_source_name(t.filter)
      sleep(1)
      print('done')
      obs_source_set_enabled(t.filter, false)
    end
    until false
  end)
end
obs_source_set_enabled(t.filter, true)
```

Thanks!

**Comment by upgradeQ** [comment]
> it would be bad to execute them all at once when OBS starts

Why so? :thinking: 
Have you tried using `accept` & `offer` ?

**Comment by tatunts** [comment]
No, sorry... I'm not very good at code, I just know a bit of the logic and copy/paste code haha

Would you mind sending me an example code that does what I need?

Send a hotkey just when the console filter is shown.

Thanks a lot!

**Comment by tatunts** [comment]
Here is an example of what I'm saying:

https://streamable.com/o9xfa4

No keypress is sent when the filter is shown, only after running it once.

If restart, same problem. Setting to auto run will trigger the script when OBS starts but after that it works (show filter), this is really bad for me, since I want to use it in about 20 sources (20x triggers at OBS start at the same time)

**Comment by upgradeQ** [comment]
Hey, no need to use 3-rd party video hosting platform, you can do it [here](https://github.blog/2021-05-13-video-uploads-available-github/) now.

**Comment by tatunts** [comment]
The video had about 500kb more than the 10mb limit, so no luck.

**Comment by tatunts** [comment]
I did try but the video had about 500kb more than the 10mb limit, so no luck.

Anyways, sorry to bother you, I'll close the issue for now.

Thanks.
